### ISSUE TITLE
Direct Component Interface

### ISSUE THREAD

**Issue by horazont** [horazont/aioxmpp#73]
A random idea of mine: Add support for using a [XEP-0114 Component](https://xmpp.org/extensions/xep-0114.html) directly. The idea is to have the component connect to an aioxmpp application. The server part of the component interface is emulated inside aioxmpp. For the application, the interface should be identical to a normal client connected to a normal server, except that the server doesn’t do S2S.

Things which need consideration:

* Roster handling
* Presence probe handling
* Handling of reconnect and authentication of the component
* possibly more

The intent is to support other protocols in an application by bundling a Transport Component (such as [biboumi](https://biboumi.louiz.org/)) and have that Transport Component connect to aioxmpp while being able to use it as if the Component was connected to a normal service.

**Comment by kousu** [comment]
Transports are interesting to me, but I am more interested in support for XEP-0114 client-side support. I am happy to run prosody or ejabberd as the real XMPP server, and hang applications (written with aioxmpp or some other platform) off of them.

Slixmpp has [`ComponentXMPP`](https://github.com/poezio/slixmpp/blob/master/slixmpp/componentxmpp.py). As far as I can tell, there is no equivalent in aioxmpp. Is that right?

**Comment by horazont** [comment]
No, there’s no equivalent in aioxmpp. The architecture of aioxmpp is at this stage rather focused around client-to-server connections. It allows us to make some choices which simplify things and allow us to provide a really high-level interface to XMPP.

The downside is that implementing components is much harder. You would have to base off XMLStream I think to get a '114 connection; not even the StanzaStream can be reused because it already assumes that it is an endpoint which can be addressed with a single JID.

**Comment by kousu** [comment]
Thank you very much for the clear explanation. Having that context makes it a lot clearer. I'll still with slixmpp for now, but good luck! I have high hopes for a solid async-first xmpp library. The way you translate XML namespaces and XEPs into python seems very tidy.
